File: impls/java/src/main/java/mal/step9_try.java
package mal;

import java.io.IOException;

import java.io.StringWriter;
import java.io.PrintWriter;
import java.util.List;
import java.util.Map;
import mal.types.*;
import mal.readline;
import mal.reader;
import mal.printer;
import mal.env.Env;
import mal.core;

public class step9_try {
    // read
    public static MalVal READ(String str) throws MalThrowable {
        return reader.read_str(str);
    }

    // eval
    public static Boolean starts_with(MalVal ast, String sym) {
        //  Liskov, forgive me
        if (ast instanceof MalList && !(ast instanceof MalVector) && ((MalList)ast).size() == 2) {
            MalVal a0 = ((MalList)ast).nth(0);
            return a0 instanceof MalSymbol && ((MalSymbol)a0).getName().equals(sym);
        }
        return false;
    }

    public static MalVal quasiquote(MalVal ast) {
        if ((ast instanceof MalSymbol || ast instanceof MalHashMap))
            return new MalList(new MalSymbol("quote"), ast);

        if (!(ast instanceof MalList))
            return ast;

        if (starts_with(ast, "unquote"))
            return ((MalList)ast).nth(1);

        MalVal res = new MalList();
        for (Integer i=((MalList)ast).size()-1; 0<=i; i--) {
            MalVal elt = ((MalList)ast).nth(i);
            if (starts_with(elt, "splice-unquote"))
                res = new MalList(new MalSymbol("concat"), ((MalList)elt).nth(1), res);
            else
                res = new MalList(new MalSymbol("cons"), quasiquote(elt), res);
        }
        if (ast instanceof MalVector)
            res = new MalList(new MalSymbol("vec"), res);
        return res;
    }

    public static MalVal EVAL(MalVal orig_ast, Env env) throws MalThrowable {
        while (true) {

        final MalVal dbgeval = env.get("DEBUG-EVAL");
        if (dbgeval != null && dbgeval != types.Nil && dbgeval != types.False)
            System.out.println("EVAL: " + printer._pr_str(orig_ast, true));

        if (orig_ast instanceof MalSymbol) {
            final String key = ((MalSymbol)orig_ast).getName();
            final MalVal val = env.get(key);
            if (val == null)
                throw new MalException("'" + key + "' not found");
            return val;
        } else if (orig_ast instanceof MalVector) {
            final MalList old_lst = (MalList)orig_ast;
            final MalVector new_lst = new MalVector();
            for (MalVal mv : (List<MalVal>)old_lst.value) {
                new_lst.conj_BANG(EVAL(mv, env));
            }
            return new_lst;
        } else if (orig_ast instanceof MalHashMap) {
            final Map<String, MalVal> old_hm = ((MalHashMap)orig_ast).value;
            MalHashMap new_hm = new MalHashMap();
            for (Map.Entry<String, MalVal> entry : old_hm.entrySet()) {
                new_hm.value.put(entry.getKey(), EVAL((MalVal)entry.getValue(), env));
            }
            return new_hm;
        } else if (!orig_ast.list_Q()) {
            return orig_ast;
        }
        final MalList ast = (MalList)orig_ast;
        MalVal a0, a1,a2, a3, res;
        // apply list
        if (ast.size() == 0) { return ast; }
        a0 = ast.nth(0);
        String a0sym = a0 instanceof MalSymbol ? ((MalSymbol)a0).getName()
                                               : "__<*fn*>__";
        switch (a0sym) {
        case "def!":
            a1 = ast.nth(1);
            a2 = ast.nth(2);
            res = EVAL(a2, env);
            env.set(((MalSymbol)a1), res);
            return res;
        case "let*":
            a1 = ast.nth(1);
            a2 = ast.nth(2);
            MalSymbol key;
            MalVal val;
            Env let_env = new Env(env);
            for(int i=0; i<((MalList)a1).size(); i+=2) {
                key = (MalSymbol)((MalList)a1).nth(i);
                val = ((MalList)a1).nth(i+1);
                let_env.set(key, EVAL(val, let_env));
            }
            orig_ast = a2;
            env = let_env;
            break;
        case "quote":
            return ast.nth(1);
        case "quasiquote":
            orig_ast = quasiquote(ast.nth(1));
            break;
        case "defmacro!":
            a1 = ast.nth(1);
            a2 = ast.nth(2);
            res = EVAL(a2, env);
            res = res.copy();
            ((MalFunction)res).setMacro();
            env.set((MalSymbol)a1, res);
            return res;
        case "try*":
            try {
                return EVAL(ast.nth(1), env);
            } catch (Throwable t) {
                if (ast.size() > 2) {
                    MalVal exc;
                    a2 = ast.nth(2);
                    MalVal a20 = ((MalList)a2).nth(0);
                    if (((MalSymbol)a20).getName().equals("catch*")) {
                        if (t instanceof MalException) {
                            exc = ((MalException)t).getValue();
                        } else {
                            StringWriter sw = new StringWriter();
                            t.printStackTrace(new PrintWriter(sw));
                            String tstr = sw.toString();
                            exc = new MalString(t.getMessage() + ": " + tstr);
                        }
                        return EVAL(((MalList)a2).nth(2),
                                    new Env(env, ((MalList)a2).slice(1,2),
                                                 new MalList(exc)));
                    }
                }
                throw t;
            }
        case "do":
            for (int i=1; i<ast.size()-1; i++)
                EVAL(ast.nth(i), env);
            orig_ast = ast.nth(ast.size()-1);
            break;
        case "if":
            a1 = ast.nth(1);
            MalVal cond = EVAL(a1, env);
            if (cond == types.Nil || cond == types.False) {
                // eval false slot form
                if (ast.size() > 3) {
                    orig_ast = ast.nth(3);
                } else {
                    return types.Nil;
                }
            } else {
                // eval true slot form
                orig_ast = ast.nth(2);
            }
            break;
        case "fn*":
            final MalList a1f = (MalList)ast.nth(1);
            final MalVal a2f = ast.nth(2);
            final Env cur_env = env;
            return new MalFunction (a2f, (mal.env.Env)env, a1f) {
                public MalVal apply(MalList args) throws MalThrowable {
                    return EVAL(a2f, new Env(cur_env, a1f, args));
                }
            };
        default:
            final MalFunction f = (MalFunction)EVAL(a0, env);
            if (f.isMacro()) {
                orig_ast = f.apply(ast.rest());
                continue;
            }
            final MalList args = new MalList();
            for (int i=1; i<ast.size(); i++)
                args.conj_BANG(EVAL(ast.nth(i), env));
            MalVal fnast = f.getAst();
            if (fnast != null) {
                orig_ast = fnast;
                env = f.genEnv(args);
            } else {
                return f.apply(args);
            }
        }

        }
    }

    // print
    public static String PRINT(MalVal exp) {
        return printer._pr_str(exp, true);
    }

    // repl
    public static MalVal RE(Env env, String str) throws MalThrowable {
        return EVAL(READ(str), env);
    }

    public static void main(String[] args) throws MalThrowable {
        String prompt = "user> ";

        final Env repl_env = new Env(null);

        // core.java: defined using Java
        for (String key : core.ns.keySet()) {
            repl_env.set(new MalSymbol(key), core.ns.get(key));
        }
        repl_env.set(new MalSymbol("eval"), new MalFunction() {
            public MalVal apply(MalList args) throws MalThrowable {
                return EVAL(args.nth(0), repl_env);
            }
        });
        MalList _argv = new MalList();
        for (Integer i=1; i < args.length; i++) {
            _argv.conj_BANG(new MalString(args[i]));
        }
        repl_env.set(new MalSymbol("*ARGV*"), _argv);


        // core.mal: defined using the language itself
        RE(repl_env, "(def! not (fn* (a) (if a false true)))");
        RE(repl_env, "(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))");
        RE(repl_env, "(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");
        
        Integer fileIdx = 0;
        if (args.length > 0 && args[0].equals("--raw")) {
            readline.mode = readline.Mode.JAVA;
            fileIdx = 1;
        }
        if (args.length > fileIdx) {
            RE(repl_env, "(load-file \"" + args[fileIdx] + "\")");
            return;
        }

        // repl loop
        while (true) {
            String line;
            try {
                line = readline.readline(prompt);
                if (line == null) { continue; }
            } catch (readline.EOFException e) {
                break;
            } catch (IOException e) {
                System.out.println("IOException: " + e.getMessage());
                break;
            }
            try {
                System.out.println(PRINT(RE(repl_env, line)));
            } catch (MalContinue e) {
            } catch (MalException e) {
                System.out.println("Error: " + printer._pr_str(e.getValue(), false));
            } catch (MalThrowable t) {
                System.out.println("Error: " + t.getMessage());
            } catch (Throwable t) {
                System.out.println("Uncaught " + t + ": " + t.getMessage());
            }
        }
    }
}

File: impls/kotlin/src/mal/step9_try.kt
package mal

import java.util.*

fun read(input: String?): MalType = read_str(input)

fun eval(_ast: MalType, _env: Env): MalType {
    var ast = _ast
    var env = _env

    while (true) {
        ast = macroexpand(ast, env)

        if (ast is MalList) {
            if (ast.count() == 0) return ast
            when ((ast.first() as? MalSymbol)?.value) {
                "def!" -> return env.set(ast.nth(1) as MalSymbol, eval(ast.nth(2), env))
                "let*" -> {
                    val childEnv = Env(env)
                    val bindings = ast.nth(1) as? ISeq ?: throw MalException("expected sequence as the first parameter to let*")

                    val it = bindings.seq().iterator()
                    while (it.hasNext()) {
                        val key = it.next()
                        if (!it.hasNext()) throw MalException("odd number of binding elements in let*")
                        childEnv.set(key as MalSymbol, eval(it.next(), childEnv))
                    }

                    env = childEnv
                    ast = ast.nth(2)
                }
                "fn*" -> return fn_STAR(ast, env)
                "do" -> {
                    eval_ast(ast.slice(1, ast.count() - 1), env)
                    ast = ast.seq().last()
                }
                "if" -> {
                    val check = eval(ast.nth(1), env)

                    if (check !== NIL && check !== FALSE) {
                        ast = ast.nth(2)
                    } else if (ast.count() > 3) {
                        ast = ast.nth(3)
                    } else return NIL
                }
                "quote" -> return ast.nth(1)
                "quasiquoteexpand" -> return quasiquote(ast.nth(1))
                "quasiquote" -> ast = quasiquote(ast.nth(1))
                "defmacro!" -> return defmacro(ast, env)
                "macroexpand" -> return macroexpand(ast.nth(1), env)
                "try*" -> return try_catch(ast, env)
                else -> {
                    val evaluated = eval_ast(ast, env) as ISeq
                    val firstEval = evaluated.first()

                    when (firstEval) {
                        is MalFnFunction -> {
                            ast = firstEval.ast
                            env = Env(firstEval.env, firstEval.params, evaluated.rest().seq())
                        }
                        is MalFunction -> return firstEval.apply(evaluated.rest())
                        else -> throw MalException("cannot execute non-function")
                    }
                }
            }
        } else return eval_ast(ast, env)
    }
}

fun eval_ast(ast: MalType, env: Env): MalType =
        when (ast) {
            is MalSymbol -> env.get(ast)
            is MalList -> ast.elements.fold(MalList(), { a, b -> a.conj_BANG(eval(b, env)); a })
            is MalVector -> ast.elements.fold(MalVector(), { a, b -> a.conj_BANG(eval(b, env)); a })
            is MalHashMap -> ast.elements.entries.fold(MalHashMap(), { a, b -> a.assoc_BANG(b.key, eval(b.value, env)); a })
            else -> ast
        }

private fun fn_STAR(ast: MalList, env: Env): MalType {
    val binds = ast.nth(1) as? ISeq ?: throw MalException("fn* requires a binding list as first parameter")
    val params = binds.seq().filterIsInstance<MalSymbol>()
    val body = ast.nth(2)

    return MalFnFunction(body, params, env, { s: ISeq -> eval(body, Env(env, params, s.seq())) })
}

private fun quasiquote(ast: MalType): MalType {
    when (ast) {
        is MalList -> {
            if (ast.count() == 2 && (ast.first() as? MalSymbol)?.value == "unquote") {
                return ast.nth(1)
            } else {
                return ast.elements.foldRight(MalList(), ::quasiquote_loop)
            }
        }
        is MalVector -> {
            val result = MalList()
            result.conj_BANG(MalSymbol("vec"))
            result.conj_BANG(ast.elements.foldRight(MalList(), ::quasiquote_loop))
            return result
        }
        is MalSymbol, is MalHashMap -> {
            val quoted = MalList()
            quoted.conj_BANG(MalSymbol("quote"))
            quoted.conj_BANG(ast)
            return quoted
        }
        else -> return ast
    }
}

private fun quasiquote_loop(elt: MalType, acc: MalList): MalList {
    val result = MalList()
    if (elt is MalList && elt.count() == 2 && (elt.first() as? MalSymbol)?.value == "splice-unquote") {
        result.conj_BANG(MalSymbol("concat"))
        result.conj_BANG(elt.nth(1))
    } else {
        result.conj_BANG(MalSymbol("cons"))
        result.conj_BANG(quasiquote(elt))
    }
    result.conj_BANG(acc)
    return result
}

private fun is_macro_call(ast: MalType, env: Env): Boolean {
    val ast_list = ast as? MalList ?: return false
    if (ast_list.count() == 0) return false
    val symbol = ast_list.first() as? MalSymbol ?: return false
    val function = env.find(symbol) as? MalFunction ?: return false

    return function.is_macro
}

private fun macroexpand(_ast: MalType, env: Env): MalType {
    var ast = _ast
    while (is_macro_call(ast, env)) {
        val symbol = (ast as MalList).first() as MalSymbol
        val function = env.find(symbol) as MalFunction
        ast = function.apply(ast.rest())
    }
    return ast
}

private fun defmacro(ast: MalList, env: Env): MalType {
    val macro = eval(ast.nth(2), env) as MalFunction
    macro.is_macro = true

    return env.set(ast.nth(1) as MalSymbol, macro)
}

private fun try_catch(ast: MalList, env: Env): MalType =
    try {
        eval(ast.nth(1), env)
    } catch (e: Exception) {
        if (ast.count() < 3) { throw e }
        val thrown = if (e is MalException) e else MalException(e.message)
        val symbol = (ast.nth(2) as MalList).nth(1) as MalSymbol

        val catchBody = (ast.nth(2) as MalList).nth(2)
        val catchEnv = Env(env)
        catchEnv.set(symbol, thrown)

        eval(catchBody, catchEnv)
    }

fun print(result: MalType) = pr_str(result, print_readably = true)

fun rep(input: String, env: Env): String =
        print(eval(read(input), env))

fun main(args: Array<String>) {
    val repl_env = Env()
    ns.forEach({ it -> repl_env.set(it.key, it.value) })

    repl_env.set(MalSymbol("*ARGV*"), MalList(args.drop(1).map({ it -> MalString(it) }).toCollection(LinkedList<MalType>())))
    repl_env.set(MalSymbol("eval"), MalFunction({ a: ISeq -> eval(a.first(), repl_env) }))

    rep("(def! not (fn* (a) (if a false true)))", repl_env)
    rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))", repl_env)
    rep("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))", repl_env)

    if (args.any()) {
        rep("(load-file \"${args[0]}\")", repl_env)
        return
    }

    while (true) {
        val input = readline("user> ")

        try {
            println(rep(input, repl_env))
        } catch (e: EofException) {
            break
        } catch (e: MalContinue) {
        } catch (e: MalException) {
            println("Error: " + e.message)
        } catch (t: Throwable) {
            println("Uncaught " + t + ": " + t.message)
            t.printStackTrace()
        }
    }
}
